src/datetime_rotation.jl

```diff
@@ -15,26 +21,35 @@ Usage example:
     logger = DatetimeRotatingFileLogger(log_dir, raw"\a\c\c\e\s\s-YYYY-mm-dd.\l\o\g")
 """
 mutable struct DatetimeRotatingFileLogger <: AbstractLogger
-    logger::SimpleLogger
+    logger::Union{SimpleLogger,FormatLogger}
     dir::String
     filename_pattern::DateFormat
     next_reopen_check::DateTime
     always_flush::Bool
 end
 
 function DatetimeRotatingFileLogger(dir, filename_pattern; always_flush=true)
-    format = DateFormat(filename_pattern)
-    return DatetimeRotatingFileLogger(
-        SimpleLogger(open(calc_logpath(dir, filename_pattern), "a"), BelowMinLevel),
-        dir,
-        format,
-        next_datetime_transition(format),
-        always_flush,
-    )
+    DatetimeRotatingFileLogger(nothing, dir, filename_pattern; always_flush=always_flush)
+end
+function DatetimeRotatingFileLogger(f::Union{Function,Nothing}, dir, filename_pattern; always_flush=true)
+    # Construct the backing logger with a temp IOBuffer that will be replaced
+    # by the correct filestream in the call to reopen! below
+    logger = if f === nothing
+        SimpleLogger(IOBuffer(), BelowMinLevel)
+    else # f isa Function
+        FormatLogger(f, IOBuffer(); always_flush=false) # no need to flush twice
+    end
+    drfl = DatetimeRotatingFileLogger(logger, abspath(dir),
+        DateFormat(filename_pattern), now(), always_flush)
+    reopen!(drfl)
+    return drfl
 end
 
+new_logger(::SimpleLogger, io) = SimpleLogger(io, BelowMinLevel)
+new_logger(l::FormatLogger, io) = FormatLogger(l.f, io, l.always_flush)
```

Comment:
perhaps `similar_logger`? Can't use `similar` since that woud be type piracy in the case of `SimpleLogger`